Fix graphical overmap notes displays & mission arrow

GoLoT opened the pull request:
#### Summary
Bugfixes "Fix graphical overmap notes displays & mission arrow"

#### Purpose of change
The new graphical overmap is not displaying the notes preview window (top-left corner) and when creating/editing notes it only displays the creation window until a key is pressed, then it disappears (while still taking inputs and working as expected, only without any feedback).

There is also an issue with the mission marker that causes errors when the map Z level is different from the mission target Z level (I think a difference >2).

#### Describe the solution
~~This PR fixes the creation note window by preventing refreshing the overmap while the note editing window is open. Previously the draw call order would cause the overmap redraw to draw on top of the window. The window refreshes automatically on keypresses, that's why the window was only displayed right after being opened and before typing anything.~~

EDIT: Previous description was slightly wrong. The issue wasn't the draw order but the sum of two different issues: There were unnecessary calls to `SDL_RenderSetClipRect()` that clipped the viewport causing rendering problems in `draw_om()` and on top of that the clipping rectangle wasn't being reset at the end of the function so the clipping bounds were affecting the draw calls for the overlaid windows. Removing the unneeded calls and adding the clipping rectangle reset at the end of `draw_om()` fixed the issues.

A new note text window is now rendered next to the currently selected tile, instead of using the old top-left corner window. This is a very basic display at the moment as I'm not good at designing UIs. Feel free to give any feedback or tweak it.

Then there is the issue of the arrow marker causing errors because of the Z level difference. Fixing the Z level when pathfinding does the trick.

EDIT: Also added fixes to align labels properly to the tile grid (Thanks @Qrox ), enabled display of paths for auto-travel and fixed NPC notes displaying for un-seen NPCs.

#### Describe alternatives you've considered
Trying to reuse the old corner window for notes. It should be as easy as creating a new window and using the same drawing code at the end of the draw_om() function. I thought that an SDL-based window would look better (if designed properly) so I went with that instead.

#### Testing
![image](https://user-images.githubusercontent.com/1606943/126880223-ce084482-d45a-4d11-9353-ffbcbd756826.png)


![image](https://user-images.githubusercontent.com/1606943/126880216-410549d6-1ae1-4cdf-b357-0149868e8c14.png)

GoLoT commented:
Pinging @KorGgenT for a review.

Qrox commented:
Other menus such as the keybindings menu and note list menu are also quirky when overmap is open. `ui_manager` is supposed to draw the UIs by the order they are created, but the overmap UI is somehow bypassing this logic.

Drawing the note next to the selected tile does look nice, but for other menus to work, the overmap UI code still needs to be fixed.

> preventing refreshing the overmap while the note editing window is open

Refreshing is required when the note editing window is open and the game window is resized.

GoLoT commented:
I tried invalidating the ui_adapter and forcing a redraw at the end of the SDL draw call: Both cases ended up in an infinite recursion as the ui_adaptor that is being invalidated is the same that does the invalidation.

I'll take a closer look at the code and see if it's possible to create a new ui_adaptor to trigger the draw calls for the SDL buffer so it can be added to the stack to control the drawing order.

GoLoT commented:
Ok, I think I have a proper solution now. I had to flag the overmap ui_adaptor as `disable_uis_below` to prevent the redraw call from invalidating windows below the overmap window (it's fullscreen anyway) and then invalidate the region for the overmap buffer.

It's working with any windows I could open from the overmap (keybinds, notes list, create note and so on). I'll do some more testing now with resizing and other interactions.

Qrox commented:
Keybindings menu is not working correctly: if you press `?` twice and press `ESC`, the first keybindings menu is open but not displayed.

Also, notes are cut off and color tags are not correctly handled:
![image](https://user-images.githubusercontent.com/11890223/126875325-473ccb45-1e60-42f5-8a10-4e34dbcc8688.png)

GoLoT commented:
I'll have to flag the keybind window as `disable_uis_below `, it seems to be the only way to prevent redrawing the overmap. Also I'll need to check the function that parses color tags to create a new one that returns `SDLcolor`s.

But before that I'll give creating an exclusive adaptor for the overmap another try. That way I might be able to control better when it should be redrawn.

Qrox commented:
I removed the city name drawing code and the keybindings menu seemed to render correctly, so there might be something wrong with city name drawing. I'll look into it.

GoLoT commented:
Hmm... there is this comment `// Labels need to be drawn last, as anything that attempts to draw a sprite after will fail.` that I assumed would be referring to the fact that `SDL_RenderSetClipRect()` is clipping the viewport and anything that he tried to draw after that didn't work, so I didn't think much else of it. If it's only the city labels that are causing issues maybe we can find a different way to draw them.

GoLoT commented:
@Qrox Can you repeat your tests after pulling the last commit? I'm trying to reproduce a problem I had where NPCs disappeared from the overmap but I'm not able to. What I noticed is that NPCs are rendered in all Z levels instead of their current level.

I'll fix that and if you can't find any other issues I'll consider the redrawing issue fixed and start working on the color tags in map notes.

Qrox commented:
I think I known what's wrong here:

When drawing the submap terrain, `SDL_RenderSetClipRect( renderer.get(), nullptr )` is called at the end of `cata_tiles::draw` to reset the clip rect, so subsequent draws are not clipped.

However, that call is missing from the `cata_tiles::draw_om`, which causes all subsequent UIs to be clipped.

When first opening a menu in the overmap UI, `ui_manager` skips the overmap drawing code and reuses the previously drawn overmap, so the clip rect is not set, and the menu is drawn correctly. When the menu is refreshed by resizing etc, though, overmap is also refreshed, so the clip rect is set, which causes all subsequent menus to be clipped and not displayed.

Resetting the clip rect at the end of `cata_tiles::draw_om` fixes all menus and NPCs are also correctly displayed, though I also notice that NPCs (and the player) are drawn on all z-levels.

GoLoT commented:
Ok, I added the missing `SDL_RenderSetClipRect( renderer.get(), nullptr )` call at the end of `draw_om()`. That should take care of any problems if `SDL_RenderSetClipRect()` calls are added. Also fixed the NPCs rendering at all Z levels. Now I think the only thing missing is the color tag parsing.

Thanks for the help.

Qrox commented on src/sdltiles.cpp:
```diff
@@ -1080,6 +1079,66 @@ void cata_tiles::draw_om( const point &dest, const tripoint_abs_omt &center_abs_
+        }
+    }
+
+    SDL_RenderSetClipRect( renderer.get(), nullptr );
```
```suggestion
    printErrorIf( SDL_RenderSetClipRect( renderer.get(), nullptr ) != 0,
                  "SDL_RenderSetClipRect failed" );
```

Qrox commented on src/sdltiles.cpp:
```diff
@@ -1050,8 +1051,6 @@ void cata_tiles::draw_om( const point &dest, const tripoint_abs_omt &center_abs_
             const int name_length = name.length();
             const point draw_pos = abs_sm_to_draw_label( pos, name_length );
             SDL_Rect clipRect = { draw_pos.x, draw_pos.y, name_length * fontwidth, fontheight };
-            printErrorIf( SDL_RenderSetClipRect( renderer.get(), &clipRect ) != 0,
```
Since this is removed, the comment above about drawing labels last can be removed.

GoLoT replied:
Done.

Qrox commented:
I've pushed a PR to your branch to fix some issues with label and note position: https://github.com/GoLoT/Cataclysm-DDA/pull/1.

GoLoT commented:
There we go. Example of colored notes:

![image](https://user-images.githubusercontent.com/1606943/126880207-41d81a64-ea91-43ab-9744-d415364b8c92.png)

logros13 commented:
don't know if you know already but while you're at it, the fast travel preview (shift+W) and map search (/) where also not working last i tried.

Qrox commented:
Fast travel preview needs to be implemented for the tiles version. Map search is fixed by this PR already.

GoLoT commented:
> don't know if you know already but while you're at it, the fast travel preview (shift+W) and map search (/) where also not working last i tried.

Added player paths using the same sprite as the cursor. Does anyone know if there is a sprite for this specifically or can think of a good one that could be used for pathing?
![image](https://user-images.githubusercontent.com/1606943/126895780-8efd2439-8ac3-4483-b428-bb0f73a405ca.png)

GoLoT commented on src/overmap_ui.cpp:
```diff
@@ -900,9 +900,11 @@ static void draw_ascii(
-    for( const auto &npc : overmap_buffer.get_npcs_near_omt( center, 0 ) ) {
-        if( !npc->marked_for_death ) {
-            corner_text.emplace_back( npc->basic_symbol_color(), npc->name );
+    if( has_debug_vision || overmap_buffer.seen( center ) ) {
```
Should this part be moved to a new PR so it can be backported?

Qrox commented:
> Added player paths using the same sprite as the cursor. Does anyone know if there is a sprite for this specifically or can think of a good one that could be used for pathing?

You can use `"highlight"`.

GoLoT commented:
> You can use `"highlight"`.

Yup, this looks much better. Thanks again.

![image](https://user-images.githubusercontent.com/1606943/126903730-44c5bf6f-9406-4446-ac6f-0a7d37f50a83.png)

KorGgenT commented on src/sdltiles.cpp:
```diff
@@ -1080,6 +1103,119 @@ void cata_tiles::draw_om( const point &dest, const tripoint_abs_omt &center_abs_
+
+        const int starting_x = draw_point.x;
+
+        for( auto &line : notes_window_text ) {
```
```suggestion
        for( std::pair<nc_color, std::string> &line : notes_window_text ) {
```

KorGgenT commented on src/sdltiles.cpp:
```diff
@@ -1080,6 +1103,119 @@ void cata_tiles::draw_om( const point &dest, const tripoint_abs_omt &center_abs_
+        const int starting_x = draw_point.x;
+
+        for( auto &line : notes_window_text ) {
+            const auto color_segments = split_by_color( line.second );
```
```suggestion
            const std::vector<std::string> color_segments = split_by_color( line.second );
```
as here

KorGgenT commented on src/sdltiles.cpp:
```diff
@@ -1080,6 +1103,119 @@ void cata_tiles::draw_om( const point &dest, const tripoint_abs_omt &center_abs_
+            draw_point.x = starting_x;
+
+            int line_length = 0;
+            for( auto seg : color_segments ) {
```
```suggestion
            for( std::string seg : color_segments ) {
```
and here

KorGgenT commented on src/sdltiles.cpp:
```diff
@@ -1080,6 +1103,119 @@ void cata_tiles::draw_om( const point &dest, const tripoint_abs_omt &center_abs_
+            geometry->rect( renderer, background_rect, SDL_Color{ 0, 0, 0, 175 } );
+
+            // Draw colored text segments
+            for( auto &colored_line : colored_lines ) {
```
```suggestion
            for( std::pair<nc_color, std::string> &colored_line : colored_lines ) {
```

Qrox commented:
Probably fixes #50529 and #49708, based on their descriptions.
